### PR TITLE
Decision Tree from Scratch 

### LINKED ISSUE #1583: Wish to add Python based scratch implementations of PCA, Decision Tree and KNN algorithms to Algo_Ds_Notes/Machine Learning
I wish to contribute the following:
1. Implementing Principal Component Analysis(PCA) from scratch
2. Implementing Decision Tree algorithm(finding best split based on entropy and Gini indexing) from scratch
3. Implementing K Nearest Neighbors(KNN) algorithm from scratch
I will also add proper visualization(if needed), documentation, code comments to explain logic.
Will also add README files

### PR DESCRIPTION
**Pull request by BALaka-18**
Fixes #1583 Decision Tree part

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Implementing the popular Decision Tree algorithm, under Supervised Machine learning, in Python
- Implementing DCT on Gini impurity concept

#### Languages Used:
- Python

#### Files Added:
- dct.py

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

File: dct.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+#tennis[tennis.columns[len(tennis.columns)-5]].value_counts()
+
+# Breaking down the calculation behind finding out the most valuable column or the first split in the decision tree
+
+# Function to find entropy of splitting node.
+def entropy(yes,no):                  # yes = probability of 'Yes' in this dataset. no = probability of 'No' in this dataset.
+    s = yes + no
+    if yes == 0:                     # If there are no 'Yes' data.
+        return -(no/s*log(no/s))
```
Comment (SakshiShreya) [inline dct.py]: Add space around operators

File: dct.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            weight = no_list[j]/sum(no_list)
+            child = child + entropy_branch*weight
+        info_gain.append(parent - child)
+    sort_index = np.argsort(np.array(info_gain)).tolist()
+    sort_index.reverse()                   # Reverse the sorted list
+    return np.array(feature_cols)[sort_index][0]     # The first column name must be the one with the highest info_gain.
+
+# Testing effectiveness on tennis dataset.
+best_split(tennis,['humidity','temp', 'wind','outlook'],'play')
+
```
Comment (SakshiShreya) [inline dct.py]: Mention sample input and output in comments at the EOF.

**Comment by somya-kapoor** [comment]
@BALaka-18 Are you working on it?

**Comment by somya-kapoor** [comment]
Due to no activity for more than a month, I am closing this issue.

**Pull request closed without merging**